test/jdk/java/net/DatagramSocket/TimeoutWithSM.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @summary Test a timed DatagramSocket.receive with a SecurityManager set
+ * @run main/othervm -Djava.security.manager=allow TimeoutWithSM
+ */
+
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.SocketTimeoutException;
+import java.security.Permission;
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class TimeoutWithSM {
+
+    private static final int TIMEOUT = 10_000;
+
+    public static void main(String[] args) throws Exception {
+        try (var socket = new DatagramSocket(null)) {
+            InetAddress lb = InetAddress.getLoopbackAddress();
+            socket.bind(new InetSocketAddress(lb, 0));
+
+            // start sender to send datagrams to us
+            var done = new AtomicBoolean();
+            SocketAddress sender = startSender(socket.getLocalSocketAddress(), done);
+
+            // set a SecurityManager that blocks datagrams from sender
+            System.setSecurityManager(new SecurityManager() {
+                @Override
+                public void checkPermission(Permission p) {
+                }
+                @Override
+                public void checkAccept(String host, int port) {
+                    var isa = new InetSocketAddress(host, port);
+                    System.out.println("checkAccept " + isa);
+                    if (isa.equals(sender)) {
+                        throw new SecurityException();
```

Comment:
Given that the goal of this test is to exercise the case where the `SecurityManager` rejects incoming packets and then times out the receive call, would it be better to just throw `SecurityException` here for all host/port combinations instead of checking the sender?

We have had cases, in the CI, where stray datagrams have been received in test cases from unexpected senders. In such cases this test will still fail, but with the assertion error that it received an unexpected packet . Throwing `SecurityException` for all addresses here, would avoid intermittent failures and in theory the test would then only get a `SocketTimeoutException`.

Reply:
> Given that the goal of this test is to exercise the case where the `SecurityManager` rejects incoming packets and then times out the receive call, would it be better to just throw `SecurityException` here for all host/port combinations instead of checking the sender?

checkAccess is called for other reasons too and the intention was just to drop datagrams from the sender. However, you may be right that there are other tests on the same host that are sending datagrams to random ports and could cause this test to fail if they happen to send it to the port that the test is using. So I'll re-test with that change.